app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskCustomView.kt

```diff
@@ -24,101 +21,107 @@ class MaskCustomView @JvmOverloads constructor(
-                                path.quadTo(mX, mY, (x + mX) / 2, (y + mY) / 2)
-
-                                mX = x
-                                mY = y
-
-                                maskLayer.onPath(path)
-                                invalidate()
-                            }
-                        }
-                    }
-                    return true
+    @SuppressLint("ClickableViewAccessibility")
+    override fun onTouchEvent(event: MotionEvent): Boolean =
+        super.onTouchEvent(event) ||
+                when (currentMode) {
+                    Mode.Zoom -> onTouchMove(event)
+                    Mode.Draw -> onTouchDraw(event)
                 }
-            }
-
-            private fun transformPointInverse(x: Float, y: Float): Pair<Float, Float> {
-                val p = floatArrayOf(x, y)
-                val inverseScaleMatrix = Matrix()
-                imageMatrix.invert(inverseScaleMatrix)
 
-                inverseScaleMatrix.mapPoints(p)
-
-                return p[0] to p[1]
+    private fun onTouchMove(event: MotionEvent) = attacher.onTouch(this, event)
+
+    private fun onTouchDraw(event: MotionEvent): Boolean {
+        val (x, y) = invert(event.x, event.y)
+        val (drawableWidth, drawableHeight) = drawableDimensions
+        val xOffBounds = 0 > x || x > drawableWidth
+        val yOffBounds = 0 > y || y > drawableHeight
+        if (xOffBounds || yOffBounds) return false
+        when (event.action) {
+            MotionEvent.ACTION_DOWN -> maskLayer.drawStart(x, y)
+            MotionEvent.ACTION_MOVE -> {
+                maskLayer.drawMove(x, y)
+                invalidate()
             }
-        })
+        }
+        return true
     }
 
-    fun getMaskBitmap(): Bitmap? {
-        return maskLayer.getBitmap()
+    private fun invert(x: Float, y: Float): Pair<Float, Float> {
+        val invertedScaleMatrix = Matrix().apply { imageMatrix.invert(this) }
+        val p = floatArrayOf(x, y).apply { invertedScaleMatrix.mapPoints(this) }
+        return p[0] to p[1]
     }
 
     override fun onDraw(canvas: Canvas) {
         super.onDraw(canvas)
-        canvas.withMatrix(imageMatrix) {
-            maskLayer.draw(this)
-        }
+        maskLayer.onDraw(canvas)
     }
 
-    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
-        super.onSizeChanged(w, h, oldw, oldh)
-        maskLayer.onViewResize(width, height)
+    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
+        super.onSizeChanged(width, height, oldWidth, oldHeight)
+        maskLayer.onViewSizeChanged(width, height)
     }
 
     override fun setImageDrawable(drawable: Drawable?) {
         super.setImageDrawable(drawable)
-
-        val drawableWidth = drawable?.intrinsicWidth ?: 0
-        val drawableHeight = drawable?.intrinsicHeight ?: 0
+        drawableDimensions = (drawable?.intrinsicWidth ?: 0) to (drawable?.intrinsicHeight ?: 0)
```

Comment:
I really like the refactorings from this class. However, I am struggling to see the benefit of this specific one, where you create a pair just to destructure it on the next line 😛